### PR TITLE
Allow cov & corr to handle missing values

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Closes https://github.com/pydata/xarray/issues/4349
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

I'm not that excited about how this is tested, very open to alternatives. And the mechanics could take another set of eyes too.

### CODE REVIEW

**Review by mathause — APPROVED** [review]
Not sure how I missed this when I reviewed the PR. Anyways, LGTM. Do you want to add a comment to the docstring that NaN values are always skipped? I'd say a `skipna` option is overkill.

**Comment by mathause** [comment]
BTW: There is a bare test with `xr.corr` that made its way into test_computation.py:

https://github.com/pydata/xarray/blob/a7fb5a9fa1a2b829181ea9e4986b959f315350dd/xarray/tests/test_computation.py#L898-L899

could you also fix this?

File: xarray/core/computation.py [hunk 1]
```diff
@@ -1256,7 +1256,9 @@ def _cov_corr(da_a, da_b, dim=None, ddof=0, method=None):
     # N.B. `skipna=False` is required or there is a bug when computing
     # auto-covariance. E.g. Try xr.cov(da,da) for
     # da = xr.DataArray([[1, 2], [1, np.nan]], dims=["x", "time"])
-    cov = (demeaned_da_a * demeaned_da_b).sum(dim=dim, skipna=False) / (valid_count)
+    cov = (demeaned_da_a * demeaned_da_b).sum(dim=dim, skipna=True, min_count=1) / (
```
[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/core/computation.py:1259]: Ah, so this is going to fail for more than one dimension. I guess the tests aren't that comprehensive. 

I think we can implement the `min_count` manually with the `valid_values`... (though potentially we could do that at the `sum` function...)

**Comment by max-sixty** [comment]
> BTW: There is a bare test with `xr.corr` that made its way into test_computation.py:
> 
> https://github.com/pydata/xarray/blob/a7fb5a9fa1a2b829181ea9e4986b959f315350dd/xarray/tests/test_computation.py#L898-L899
> 
> could you also fix this?

Sure — though what's wrong with it?

File: xarray/tests/test_computation.py [hunk 2]
```diff
@@ -862,8 +862,10 @@ def np_cov_ind(ts1, ts2, a, x):
             ts1, ts2 = broadcast(ts1, ts2)
             valid_values = ts1.notnull() & ts2.notnull()
 
-            ts1 = ts1.where(valid_values)
-            ts2 = ts2.where(valid_values)
+            # While dropping isn't ideal here, numpy will return nan
+            # if any segment contains a NaN.
+            ts1 = ts1.where(valid_values, drop=True)
```
[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/tests/test_computation.py:867]: I'm not so familiar with numpy's masking. Will that work?

```python

In [8]: import numpy.ma as ma
   ...:
   ...: a = np.arange(5, dtype=float)
   ...:
   ...: a[2] = np.NaN
   ...:
   ...: a[3] = np.PINF
   ...:
   ...: a
Out[8]: array([ 0.,  1., nan, inf,  4.])

In [9]: ma.masked_invalid(a)
Out[9]:
masked_array(data=[0.0, 1.0, --, --, 4.0],
             mask=[False, False,  True,  True, False],
       fill_value=1e+20)

In [10]: np.cov(_,_)
Out[10]:
array([[nan, nan],
       [nan, nan]])
```

**Comment by mathause** [comment]
The test is not in a function. I think the link to the test looks like it is indented but it's not...

File: xarray/tests/test_computation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mathause) [reply xarray/tests/test_computation.py:867]: You need `np.ma.cov(...)`. However, we don't skip inf (`xr.DataArray([np.inf]).isnull()` is `False`) but `masked_invalid` does. Thus, `np.cov(masked_a, masked_a)` will probably not evaluate to the same as your fixed `xr.cov(da_a, da_a)`.

[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/tests/test_computation.py:868]: Ah great, but `inf` aside, it works! I reckon this is better, I'll make the change

File: xarray/core/computation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/core/computation.py:1259]: Thanks for adding https://github.com/pydata/xarray/issues/4354

**Comment by AndrewWilliams3142** [comment]
> The test is not in a function. I think the link to the test looks like it is indented but it's not...

My bad ;)

**Comment by max-sixty** [comment]
> My bad ;)

Ha, no stress @AndrewWilliams3142 !

File: xarray/tests/test_computation.py [hunk 3]
```diff
@@ -818,8 +818,9 @@ def test_vectorize_dask():
     assert_identical(expected, actual)
 
 
-with raises_regex(TypeError, "Only xr.DataArray is supported"):
-    xr.corr(xr.Dataset(), xr.Dataset())
+def test_oorr_only_dataarray():
```
Comment (keewis) [inline xarray/tests/test_computation.py:821]:
```suggestion
def test_corr_only_dataarray():
```

**Comment by max-sixty** [comment]
This is ready to go on green

**Comment by dcherian** [comment]
Thanks @max-sixty

**Pull request merged**